lib/ota/portable/nordic/nrf52840-dk/aws_ota_pal.c

```diff
@@ -520,33 +520,29 @@ ret_code_t prvWriteFlash( uint32_t ulOffset,
                           uint32_t ulBlockSize,
                           uint8_t * const pacData )
 {
-    uint32_t ul32BitBlocksTosend, ulTotal32BitBlocks, ul32BitBlocksSent;
+    uint32_t ulByteSent, ulByteToSend;
     ret_code_t xErrCode = NRF_SUCCESS;
-    ulTotal32BitBlocks = (ulBlockSize+3)/4;
-    ul32BitBlocksSent = 0;
 
-
-    while(ul32BitBlocksSent < ulTotal32BitBlocks)
+    ulByteSent = 0;
+    while(ulByteSent < ulBlockSize)
     {
-      ul32BitBlocksTosend = 0;
-
-      ul32BitBlocksTosend = ulTotal32BitBlocks - ul32BitBlocksSent;
+      ulByteToSend = ulBlockSize - ulByteSent;
 
-      if(ul32BitBlocksTosend > otapalSERIALIZING_ARRAY_SIZE)
+      if(ulByteToSend > otapalSERIALIZING_ARRAY_SIZE*4)
       {
-        ul32BitBlocksTosend = otapalSERIALIZING_ARRAY_SIZE;
+        ulByteToSend = otapalSERIALIZING_ARRAY_SIZE*4;
       }
 
-      if(ul32BitBlocksTosend > NRF_FICR->CODEPAGESIZE)
+      if(ulByteToSend > NRF_FICR->CODEPAGESIZE)
       {
-        ul32BitBlocksTosend = NRF_FICR->CODEPAGESIZE;
+        ulByteToSend = NRF_FICR->CODEPAGESIZE*4;
       }
 
-      memcpy(pulSerializingArray, pacData + (ul32BitBlocksSent*4), ul32BitBlocksTosend); 
+      memcpy(pulSerializingArray, pacData + ulByteSent, ulByteToSend); 
 
       xEventGroupClearBits( xFlashEventGrp, otapalFLASH_SUCCESS | otapalFLASH_FAILURE );
       /* Softdevice can write only by 32-bit words */
-      ret_code_t xErrCode = sd_flash_write( ( uint32_t * ) (ulOffset + ul32BitBlocksSent), pulSerializingArray, ul32BitBlocksTosend );
+      ret_code_t xErrCode = sd_flash_write( ( uint32_t * ) (ulOffset + ulByteSent), pulSerializingArray, (ulByteToSend+3)/4);
```

Comment:
Reset pulSerializeArray to zero, as it writes number of bytes rounded up to next power of 4, it could write some garbage value for the last iteration.

Reply:
It is fine, it can write garbage, it wont be used. block is garanteed to be a power of 4